### ISSUE TITLE
drag tab outside should be result into an new window :)

### ISSUE THREAD

**Issue by boonkerz** [Adanos020/egui_dock#4]
nice to have i think :)

link maybe help: 
https://github.com/vivlim/egui-glow-multiwin

**Comment by Adanos020** [comment]
Agree, this feature is going to be crucial to get the library to a usable state

**Comment by Jerrody** [comment]
It can be done with:
https://docs.rs/egui/latest/egui/containers/struct.Window.html
```rs
egui::Window::new("My Window").show(ctx, |ui| {
   ui.label("Hello World!");
});
```

**Comment by Adanos020** [comment]
One way of achieving this effect is to imitate Dear Imgui.

![image](https://user-images.githubusercontent.com/2311185/184452547-50ce2945-ec88-412d-b9bb-513fbd3f507f.png)

While the tab is being dragged outside of the tab bar, it is a new window until you drop it while hovering over one of these icons that show up inside the node you're currently hovering over. If it's dropped anywhere else, it stays as a window.

We might consider implementing this for all `egui::Window`s or creating our custom `egui_dock::DockableWindow` (or whatever we're gonna call it) instead.

**Comment by Jerrody** [comment]
> One way of achieving this effect is to imitate Dear Imgui.
> 
> ![image](https://user-images.githubusercontent.com/2311185/184452547-50ce2945-ec88-412d-b9bb-513fbd3f507f.png)
> 
> While the tab is being dragged outside of the tab bar, it is a new window until you drop it while hovering over one of these icons that show up in the middle of the node you're currently hovering over. If it's dropped anywhere else, it stays as a window.
> 
> We might consider implementing this for all `egui::Window`s or creating our custom `egui_dock::DockableWindow` (or whatever we're gonna call it) instead.

If we are going to do this it will require big changes.

**Comment by Adanos020** [comment]
Definitely. But I think it's going to be worth it as it would make using the docking UI much easier.

Of course, I'm open to other ideas too, this one's just the first thing that came into my mind.

**Comment by Jerrody** [comment]
> Definitely. But I think it's going to be worth it as it would make using the docking UI much easier.
> 
> Of course, I'm open to other ideas too, this one's just the first thing that came into my mind.

Or (it is simple and seems pretty elegant) it's just right click on tab and then this tab becomes individual window.

**Comment by Jerrody** [comment]
> One way of achieving this effect is to imitate Dear Imgui.
> 
> ![image](https://user-images.githubusercontent.com/2311185/184452547-50ce2945-ec88-412d-b9bb-513fbd3f507f.png)
> 
> While the tab is being dragged outside of the tab bar, it is a new window until you drop it while hovering over one of these icons that show up in the middle of the node you're currently hovering over. If it's dropped anywhere else, it stays as a window.
> 
> We might consider implementing this for all `egui::Window`s or creating our custom `egui_dock::DockableWindow` (or whatever we're gonna call it) instead.

I have an idea. If to make this area remove it and replace it with creating new window.
![image](https://user-images.githubusercontent.com/68968657/184454339-6e5b4c12-0cd8-4a40-98a8-ba05c34c25d9.png)

**Comment by Adanos020** [comment]
> Or (it is simple and seems pretty elegant) it's just right-click on a tab and then this tab becomes an individual window.

I have a couple problems with this idea:
- It's not nearly as intuitive to the user as pulling the tab out of the tab bar
- How would you dock the window back into a specific node?

**Comment by Adanos020** [comment]
> I have an idea. What if to remove this area and replace it with creating a new window? ![image](https://user-images.githubusercontent.com/68968657/184454339-6e5b4c12-0cd8-4a40-98a8-ba05c34c25d9.png)

Can you elaborate more on the exact behaviour you have in mind? I'm not quite sure what area you mean to remove, among other things.

**Comment by Jerrody** [comment]
> > I have an idea. What if to remove this area and replace it with creating a new window? ![image](https://user-images.githubusercontent.com/68968657/184454339-6e5b4c12-0cd8-4a40-98a8-ba05c34c25d9.png)
> 
> Can you elaborate more on the exact behaviour you have in mind? I'm not quite sure what area you mean to remove, among other things.

I mean instead of showing this selection when the user drops a tab and there's no selection it will create an individual winow. And it doesn't require global changes.

**Comment by Jerrody** [comment]
Okay, I determined how to do this.
```rs
                println!("{:?} {:?}", helper.size(), hover.size());
```

**Comment by Jerrody** [comment]
> > > I have an idea. What if to remove this area and replace it with creating a new window? ![image](https://user-images.githubusercontent.com/68968657/184454339-6e5b4c12-0cd8-4a40-98a8-ba05c34c25d9.png)
> > 
> > 
> > Can you elaborate more on the exact behaviour you have in mind? I'm not quite sure what area you mean to remove, among other things.
> 
> I mean instead of showing this selection when the user drops a tab and there's no selection it will create an individual winow. And it doesn't require global changes. Need just understand that area fully filled.

When selection fills area of node.

**Comment by Jerrody** [comment]
Sorry, if I explain pretty odd.

**Comment by Adanos020** [comment]
So if I understand correctly: dragging the tab looks like it does right now and dropping it in a certain spot should create a new window out of it, or dock it in a node if it's dropped in another certain spot. Is that right?

I'm not yet sure how it would feel to the user, but it's worth giving it a try!

**Comment by Jerrody** [comment]
> So if I understand correctly: dragging the tab looks like it does right now and dropping it in a certain spot should create a new window out of it, or dock it in a node if it's dropped in another certain spot. Is that right?
> 
> I'm not yet sure how it would feel to the user, but it's worth giving it a try!

My current state of experiments.
![ezgif com-gif-maker](https://user-images.githubusercontent.com/68968657/184496483-1a8fedd7-9527-489d-a1e7-46dbdcdfb01b.gif)

**Comment by Jerrody** [comment]
It works pretty good other thing is how to translate content to the window and how to back this window.

**Comment by Jerrody** [comment]
I think that should be great idea if to create a `nightly` branch where we can do experiments and others dark things :')

**Comment by Adanos020** [comment]
Good progress so far!

> I think that should be great idea if to create a `nightly` branch where we can do experiments and others dark things :')

That's not a bad idea, I'll create one now. We can also use draft pull requests for tracking all WIP stuff.

**Comment by Jerrody** [comment]
> That's not a bad idea, I'll create one now. We can also use draft pull requests for tracking all WIP stuff.

Sounds good.

**Comment by Adanos020** [comment]
Done, you can use the [experimental](https://github.com/Adanos020/egui_dock/tree/experimental) branch :)

**Comment by smallB007** [comment]
Guys, if I may add to that discussion. If anyone wants to know how docking should behave, please look how Visual Studio (not VSCode). They (Visual Studio) simply perfected it.

**Comment by Adanos020** [comment]
> Guys, if I may add to that discussion. If anyone wants to know how docking should behave, please look how Visual Studio (not VSCode). They (Visual Studio) simply perfected it.

Yes, this is the approach I would like to go with, that's why I'm suggesting to imitate Dear ImGui.

**Comment by Adanos020** [comment]
I created a discussion for this feature: https://github.com/Adanos020/egui_dock/discussions/29

**Comment by gre-v-el** [comment]
> nice to have i think :)
> 
> link maybe help: 
> https://github.com/vivlim/egui-glow-multiwin
I know this issue is pretty old, but judging by the linked repo I think the original poster meant the os window, not the `egui::Window` in their feature request. See the linked YouTube video in the README of the linked repo.

**Comment by micfong-z** [comment]
Upon some tweaking (with breaking changes), I'm able to make it this far (code at https://github.com/micfong-z/egui_dock/tree/main; definitely not ready for a PR), with a imgui-like overlay:

https://github.com/Adanos020/egui_dock/assets/68492819/1e04a70a-b82e-4bb5-9743-0039425ab096

I'm using a separate `Vec<(Tab, Pos2)>` in `DockArea` to store all detached windows (`Pos2` used to track init position of window).

However, same as in https://github.com/Adanos020/egui_dock/pull/27, I have totally no idea how to handle the windows after they are "detached".

Maybe we should make `DockArea` contain a `Vec<Tree<Tab>>` so that all detached windows have individual `Tree`s to handle the tabs in it, and the tabs *should* be able to interact with other `Tree`s on other windows. (AFAIK, imgui gives all detached windows new tab managing areas)

*I'm relatively new to Rust, and so its data structures. So there might be better solutions (?)*

**Comment by Adanos020** [comment]
@micfong-z Great work, this looks incredible!

I think the issue with detached windows might be that we'd need to make changes to egui itself, so that we could "steal" the contents of any `egui::Window`, close it, and put them in a tab. Another option is to create our own `DockableWindow` and capture its instances.

It's not an easy problem to solve and will probably take a few iterations to get this right.

I also think it might be worthwhile to create a new branch just for this feature. You could then open a PR to that branch so that other people would also be able to contribute.

**Comment by Adanos020** [comment]
Closed by #149
